### PR TITLE
Adds observer api for closing connection event

### PR DESCRIPTION
**Pull request by ayush-yadav001**
Add connection on shutdown observer if a connection was closed due to any reason.

### Motivation:

Allow client to react on connection close events. Once we connect a socket channel connection successfully, we need an api to observe if connection closed at any point of time .

For example client can handle reconnects if connection was closed

### Modifications:

Added onShutdown() method to observe closing event in ReactiveSwiftClient.swift file.
Added testConnectionDisposeListener() test in RSocketReactiveSwiftTests.swift file.

### Result:

We can now easily observe closing event.

### CODE REVIEW

File: Tests/RSocketReactiveSwiftTests/RSocketReactiveSwiftTests.swift [hunk 1]
```diff
@@ -404,4 +404,46 @@ final class RSocketReactiveSwiftTests: XCTestCase {
...
+        XCTAssertNotNil(client)
+        // client connection closed event signal producer
+        client.shutdownProducer.startWithSignal({ signal, interruptHandle in
+            signal.flatMapError({ error -> Signal<Void, Error> in
+                XCTFail("\(error)")
+                return .empty
+            }).materialize().collect().observeValues { values in
+                didReceiveConnectionclosedEvent.fulfill()
+            }
+        })
```
Comment (nkristek) [inline Tests/RSocketReactiveSwiftTests/RSocketReactiveSwiftTests.swift:439]: Nit: You could also start the producer directly and check for the `completed` event:
```swift
client.shutdownProducer.start {
    switch $0 {
    case .completed:
        didReceiveConnectionclosedEvent.fulfill()
    default: break
    }
}
```

File: Sources/RSocketReactiveSwift/Client/ReactiveSwiftClient.swift [hunk 2]
```diff
@@ -46,6 +46,20 @@ public struct ReactiveSwiftClient: Client {
...
     }
+    /// This methods helps to get call back whenever connection is closed
+    /// - Returns: SignalProducer<Void, Swift.Error> to represent task result
+    public var shutdownProducer: SignalProducer<Void, Swift.Error> {
+        SignalProducer { observer, _ in
+            coreClient.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    observer.sendCompleted()
+                case.failure(let error):
```
Comment (nkristek) [inline Sources/RSocketReactiveSwift/Client/ReactiveSwiftClient.swift:57]:
```suggestion
                case .failure(let error):
```

**Review by nkristek — APPROVED** [review]
LGTM

File: Tests/RSocketReactiveSwiftTests/RSocketReactiveSwiftTests.swift (hunk 1 above)
Comment (ayush-yadav001) [reply Tests/RSocketReactiveSwiftTests/RSocketReactiveSwiftTests.swift:439]: sure will consider this for future reference , looks more appropriate , Thanks.

**Review by OlegDokuka — APPROVED** [review]

**Pull request merged**